passes/cmds/partition.cc

```diff
@@ -0,0 +1,501 @@
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/cost.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct PartitionPass : public Pass {
+	//Constructs a hypergraph from a flat techmapped module, partitions the hypergraph,
+	//recovers the cell partition assignments, creates partition modules with the appropriate
+	//cells, ports, and intra-partition connections, then replaces the original module
+	//with one that instantiates and connects the partition modules.
+	int opt_k, opt_s, opt_imbalance;
+	bool opt_verbose, opt_nocleanup;
+	RTLIL::Module *module, *new_module;
+	RTLIL::Design *design;
+	dict<RTLIL::IdString, int> cell_type_costs;
+	dict<RTLIL::SigBit, int> sigbit_to_edgenum;
+	dict<int, pool<int>> edge_nodes;
+	dict<int, pool<int>> partition_nodes;
+	dict<int, bool> edge_cut;
+	pool<RTLIL::IdString> cut_wires;
+	int node_ctr, edge_ctr, cut_edge_ctr, partitions;
+	std::vector<RTLIL::Module *> partition_modules;
+
+	PartitionPass() : Pass("partition", "split techmapped module into equal-sized parts"), opt_k(-1), opt_s(-1),
+	                  opt_imbalance(5), opt_verbose(false), opt_nocleanup(false), module(nullptr), new_module(nullptr),
+	                  design(nullptr), cell_type_costs(CellCosts::unit_gate_cost()), node_ctr(0), edge_ctr(0),
+	                  cut_edge_ctr(0), partitions(2) { }
+
+	void reset() {
+		opt_k = -1, opt_s = -1, opt_imbalance = 5;
+		opt_verbose = false, opt_nocleanup = false;
+		module = nullptr, new_module = nullptr;
+		design = nullptr;
+		cell_type_costs = CellCosts::unit_gate_cost();
+		sigbit_to_edgenum.clear();
+		edge_nodes.clear();
+		partition_nodes.clear();
+		edge_cut.clear();
+		cut_wires.clear();
+		node_ctr = 0, edge_ctr = 0, cut_edge_ctr = 0, partitions = 2;
+		partition_modules.clear();
+	}
+
+	void create_hypergraph() {
+		//Initial set up on sigbit_to_edgenum for the cell output ports and on edge_nodes with just that output node for now:
+		for (auto cell : module->cells()) {
+			if (cell_type_costs.find(cell->type) == cell_type_costs.end())
+				log_cmd_error("Module must be techmapped; encountered unknown cell type %s\n", cell->type.c_str());
+
+			if (opt_verbose) log("Adding cell %s as node #%d\n", cell->name.c_str(), node_ctr);
+			for (auto port : cell->connections()) {
+				if (cell->output(port.first)) {
```

Comment:
The amount of rightward drift and nested conditionals in this code (here specifically and everywhere below) makes it excessively hard to review. After looking at this PR I find myself in strong agreement with [Claire's view](https://github.com/YosysHQ/yosys/pull/2154#issuecomment-646136574).

Reply:
I think the problem is really GitHub's crappy UI.  This PR looks just fine to me even on a 1368x768 laptop screen with tabstop=4.

But if you're really insistent, I can obfuscate the control flow with `continue;` instead of `if()` guards etc.

Reply:
Tabstops solve a part of the problem, but I consider nested conditionals obfuscation, not early exits.

Reply:
I understand there's legitimate room for disagreement, and you seem insistent, so I'll change it.

Reply:
Thank you for accomodating!